File: 3-leetcode/01-easy/js/108-convert-sorted-array-to-binary-search-tree.js
const sortedArrayToBST = function (nums) {
  let left = 0;
  let right = nums.length - 1;
  let mid = Math.floor(left + right / 2);
  let head = new TreeNode(
    nums[mid],
    recursive(nums, left, mid - 1),
    recursive(nums, mid + 1, right)
  );

  function recursive(arr, left = 0, right = nums.length - 1) {
    // base case
    if (left > right) return;

    let mid = Math.floor((left + right) / 2);
    let node = new TreeNode(
      nums[mid],
      recursive(nums, left, mid - 1),
      recursive(nums, mid + 1, right)
    );

    return node;
  }
  return head;
};

File: 3-leetcode/01-easy/cpp/108-convert-sorted-array-to-binary-search-tree.cpp
#include <bits/stdc++.h>

using namespace std;

struct TreeNode
{
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
  TreeNode *recur(vector<int> &arr, int left, int right)
  {
    if (left > right)
    {
      return NULL;
    }
    int mid = (left + right) / 2;
    TreeNode *node = new TreeNode(arr[mid], recur(arr, left, mid - 1), recur(arr, mid + 1, right));

    return node;
  }

  TreeNode *sortedArrayToBST(vector<int> &nums)
  {
    int left = 0, right = nums.size() - 1;
    int mid = (left + right) / 2;
    TreeNode *head = new TreeNode(nums[mid], recur(nums, left, mid - 1),
                                  recur(nums, mid + 1, right));
    return head;
  }
};
